Added Support for Events on Audio Play Completion

CameronVetter opened the pull request:
Implements enhancement #116

msftclas commented:
Hi __@CameronVetter__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@CameronVetter__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

NeerajW commented:
Looks good to me. @joekellyms any comments on this PR?

joekellyms commented:
This is great! I just have a couple of small suggestions:

1. I think the name "messageToSend" is a little unclear/misleading. It should probably denote being something that happens when the sound completes, and also the fact that it's another event - not to be confused with another type of "message". So maybe something like "completionEvent"?
2. Calling this event might be best done inside of the RemoveEventInstance function. That way, it can also be sent when you stop an event - like a looping sound, for instance.

Thoughts?

NeerajW commented on Assets/HoloToolkit/SpatialSound/Scripts/UAudioManager/ActiveEvent.cs:
```diff
@@ -61,6 +61,12 @@ public GameObject AudioEmitter
             private set;
         }
 
+        public string MessageToSend
```
I also wonder if you make this into an actual event that people can sign up for?

CameronVetter replied:
I did it using unity messages to follow the unity model for this sort of thing, which I've noticed the rest of the HoloToolkit does.  Examples of the top of my head are "OnGazeEntered", "OnSelect", but know there are a bunch more.  Are there any examples in the HoloToolkit where someone follows the .net event pattern instead of the unity message pattern?

stbertou replied:
What about using a proper unit event ? (see OnGazeEnterEvent for ex.)
That sounds cleaner to me instead of using string based messages.
Just my 2c.

jwittner replied:
I was thinking the same thing, but wasn't sure how ActiveEvent presented itself in the editor. If it's configurable in the editor I also think that the UnityEvent is the way to go!

stbertou replied:
Performance wise, UnityEvent are slower than delegates but they can nicely be set straight from the inspector so I find them very powerful.
It's also cool how you create your own dynamic events (for ex. call this function that accept a Color), so to me it's definitely the way to go.
To finish we all know how slow SendMessage is ! ;-)

CameronVetter replied:
I'm fairly new to Unity, and I believe I'm following the conversation here, but I'm go to restate it just to make sure. So please feel free to aim me in the right direction if I've got something wrong.  I used the SendMessage Event system which is baked into Unity to do this, and I understand it is quite slow because it is using reflection internally.  

From reading above I believe that @stbertou and @NeerajW are suggesting the same thing, which is to follow the pattern recently added for OnGazeEnterEvent.  The advantage of this is that I get visibility in the Unity editor and can configure the receiver of the event there instead of passing it on the play method as I currently have it implemented.  This does not result in an improvement in performance since it still uses SendMessage to decide when to trigger the event.

Here is why I don't think this is appropriate for this situation.  Imagine that I have UAudioManager configure to play 3 sounds: Sound1, Sound2, Sound3.  Now imagine that I want  to do something after Sound1 is complete but not after Sound2 and Sound3.  How I currently have this implemented I would do this with:

```
audio.PlayEvent("Sound1", gameObject, "OnSound1Complete");
audio.PlayEvent("Sound2", gameObject);
audio.PlayEvent("Sound3", gameObject);
```

Following the unity event pattern I don't see how you accomplish the same thing, you would attach an event  and it would get executed after every sound is completed. 

Does that make sense?

jwittner replied:
Right, I think that's in line with my question about ActiveEvent being present in the editor. It seems like right now only the AudioSource would be present in the editor right?

CameronVetter replied:
That is correct AudioSource is present in the editor but ActiveEvent is not.  There is no change in this behavior.

CameronVetter commented:
Those are both good points
1. I changed the property name to "MessageOnAudioEnd"
2. You are right this does belong in RemoveEventInstance, good catch.

I've updated the PR to reflect both changes.

jwittner commented:
Regarding the use of UnityEvent, what about having both options?

jwittner commented:
I'm down to pull this in, anything holding us back still @NeerajW?

CameronVetter commented:
Is there anything I can do to help get this merged in?

ChangweiZhang commented:
I have made a different solution:
`StartCoroutine(PlaySoundWithCallback());`

` IEnumerator PlaySoundWithCallback()
    {

        UAudioManager.Instance.PlayEvent("_Welcome", audioSource);
        yield return new WaitForSeconds(audioSource.clip.length);
        //do something
   }`

CameronVetter commented:
@ChangweiZhang Unfortunately that doesn't work in all cases.  If something else cancels the playing of the sound prior to its completion, the event will not go off right away.  Also if the UAudioManager contains a delay prior to the sound being played or multiple sounds played in sequence this will not work.
